R/detect_mistakes.R

```diff
@@ -1,281 +1,302 @@
+    remaining_solution_names <- solution_names[solution_names != ""]
+    
+    ## Do any solution names partially match multiple user names?
+    pmatches_per_formal <- function(solution_name) {
+      sum(startsWith(solution_name, remaining_user_names))
+    }
+    
+    matches <- vapply(remaining_solution_names, pmatches_per_formal, 1)
+    offenders <- matches[matches > 1]
+    
+    if (length(offenders) > 0) {
+      overmatched_name <- rlang::names2(offenders[1])
+      return(
+        too_many_matches(
+          this_call = user, 
+          that_name = overmatched_name,
+          enclosing_call = enclosing_call,
+          enclosing_arg = enclosing_arg
         )
-      }
+      )
     }
 
-  } else if (user_len < solution_len) {
-    for (i in seq_len(solution_len)) {
-      if (i > user_len || user_names[i] != solution_names[i]) {
-        return(
-          missing_argument(
-            this_call = solution,
-            that = solution[[i]],
-            that_name = solution_names[[i]]
-          )
+    ## How many formals does each user arg partially match?
+    pmatches_per_arg <- function(user_name) {
+      sum(startsWith(remaining_solution_names, user_name))
+    }
+    
+    matches <- vapply(remaining_user_names, pmatches_per_arg, 1)
+    offenders <- matches[matches > 1]
+    unused <- matches[matches == 0]
+    
+    # names that match multiple arguments are a syntax error
+    if (length(offenders) > 0) {
+      bad_name <- rlang::names2(offenders[1])
+      return(
+        bad_argument_name(
+          this_call = user, 
+          this = user[[bad_name]], 
+          this_name = bad_name,
+          enclosing_call = enclosing_call,
+          enclosing_arg = enclosing_arg
         )
-      }
+      )
     }
-  } else {
-    # see if call is same
-    if (!identical(user[[1]], solution[[1]])) {
-      return(wrong_value(this = deparse_to_string(user),
-                         that = prep(solution),
-                         this_name = user_names[1],
-                         that_name = solution_names[1]))
+    
+    # Unmatched named arguments are surplus
+    if (length(unused) > 0) {
+      surplus_name <- rlang::names2(unused[1])
+      return(
+        surplus_argument(
+          this_call = user,
+          this = user[[surplus_name]],
+          this_name = surplus_name,
+          enclosing_call = enclosing_call,
+          enclosing_arg = enclosing_arg
+        )
+      )
     }
+  }
     
-    for (i in seq_len(user_len)) {
-      if (user_names[i] != solution_names[i]) {
-         return(
-           surplus_argument(
-             this_call = user,
-             this = user[[i]],
-             this_name = user_names[[i]]
-           )
-         )
-      }
-      # if user[i] solution[i] not same: isolate_mismatch_2
-      if (!identical(user[[i]], solution[[i]])) {
-        return(
-          detect_mistakes(user[[i]], solution[[i]], env = env)
+  # Check for unused unnamed arguments
```

Comment:
Add comment about how all user args will be unnamed arguments from this point on